src/macaron/slsa_analyzer/provenance/loader.py

```diff
@@ -103,9 +130,90 @@ def _load_provenance_file_content(
     if not isinstance(json_payload, dict):
         raise LoadIntotoAttestationError("The provenance payload is not a JSON object.")
 
+    if json_payload["predicate"]:
+        return json_payload
+
+    # For provenance without a predicate (e.g. PyPI), try to use the provenance certificate instead.
+    raw_certificate = json_extract(provenance, ["verification_material", "certificate"], str)
+    if not raw_certificate:
+        raise LoadIntotoAttestationError("Failed to extract certificate data.")
+    try:
+        decoded_certificate = base64.b64decode(raw_certificate)
+        certificate_claims = get_x509_certificate_values(decoded_certificate)
+    except UnicodeDecodeError as error:
+        raise LoadIntotoAttestationError("Cannot decode the payload.") from error
+    except ValueError as error:
+        logger.debug(error)
+        raise LoadIntotoAttestationError("Error parsing certificate.") from error
+
+    pypi_predicate = PyPICertificatePredicate.build_predicate(
+        certificate_claims["source_repo"],
+        certificate_claims["source_digest"],
+        certificate_claims["workflow"],
+        certificate_claims["invocation"],
+    )
+    json_payload["predicate"] = pypi_predicate
     return json_payload
 
 
+def get_x509_certificate_values(decoded_certificate: bytes) -> dict:
+    """Retrieve the values of interest from an x509 certificate.
+
+    Parameters
+    ----------
+    decoded_certificate: bytes
+        The decoded certificate bytes.
+
+    Returns
+    -------
+    dict
+        A dictionary of the extracted values.
+
+    Raises
+    ------
+    ValueError
+        If the values could not be extracted.
+    """
+    certificate = x509.load_der_x509_certificate(decoded_certificate)
+    certificate_claims = {}
+    for extension in certificate.extensions:
+        if extension.oid.dotted_string not in _OID_IDS:
+            continue
+
+        claim_name = _OID_IDS[extension.oid.dotted_string]
+        certificate_claims[claim_name] = extension.value.value
+
+    for name in _OID_NAMES:
+        if name not in certificate_claims:
+            raise ValueError(f"Missing certificate value: {name}")
+
+        # Values are DER encoded UTF-8 strings. Removing the first two bytes seems to be sufficient.
+        value: str = certificate_claims[name][2:].decode("UTF-8")
```

Comment:
Could we perform this operation in the first loop ?
```
for extension in certificate.extensions:
    ...
```
I think the purpose of this loop is for validating `certificate_claims` content, so in my opinion, it's best if we don't do extra mutation.

Reply:
Thanks. It looks okay to me. Just a small observation that the logic for checking whether `certificate_claims` contains enough claim names (e.g. source_repo, source_digest, etc.) has been removed in the latest commit. Just want to make sure that this is intended.

Reply:
Good catch. I'll add it back in.